### ISSUE TITLE
JSON Protocol Proposal

### ISSUE THREAD

**Issue by kRITZCREEK** [kRITZCREEK/psc-ide#43]
I sat down for a while today and thought of a new JSON based API for querying psc-ide. The general idea is to keep the API extensible so we can iterate on it.

The reason for the separation in between Filters and Matchers is that I want to support use cases like searching for [Levenshtein Distance](https://en.wikipedia.org/wiki/Levenshtein_distance) which can suggest likely corrections in case of a typo.

So Filters only know True or False and Matchers can score their results more gradual. 

### Command:
```json
{
  "command": "load",
  "params": {
    "modules": ["Module.Name1", "Module.Name2"],
    "dependencies": ["Module.Name3"]
  }
}

{
  "command": "type",
  "params": {
    "search": "filterM",
    "filters": [Filter]
  }
}

{
  "command": "complete",
  "params": {
    "filters": [Filter],
    "matcher": (optional) Matcher
  }
}

{
  "command": "list|cwd|quit"
}
```
### Matcher:
```json

{
  "matcher": "flex",
  "params": {
    "search": "filt"
  }
}

{
  "matcher": "helm",
  "params": {
    "searches": ["fi", "ter"]
  }
}

{
  "matcher": "distance",
  "params": {
    "search": "dilterM",
    "maxDist": 3
  }
}
```
### Filter:
```json
{
  "filter": "exact",
  "params": {
    "search": "filterM"
  }
}

{
   "filter": "prefix",
   "params": {
     "search": "filt"
   }
}

{
   "filter": "modules",
   "params": {
     "modules": ["My.Module"]
   }
}

{
  "filter": "dependencies",
  "params": {
    "modules": ["My.Module"]
  }
}

```

**Comment by kRITZCREEK** [comment]
ping @epost @nwolverson @dbushenko

**Comment by nwolverson** [comment]
No particular comment, just watch for spending time making fancy stuff that might not be necessary. Unless it's fun.

Your list of filters, I guess that is `AND`? Might be useful to note optional fields if any.

**Comment by kRITZCREEK** [comment]
implemented in #48
